src/persist-client/src/read.rs

```diff
@@ -0,0 +1,137 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::marker::PhantomData;
+use std::num::NonZeroUsize;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+use differential_dataflow::difference::Semigroup;
+use futures_util::Stream;
+use mz_persist_types::{Codec, Codec64};
+use serde::{Deserialize, Serialize};
+use timely::progress::{Antichain, Timestamp};
+use tracing::warn;
+
+use crate::error::Error;
+
+// WIP This probably doesn't need to be Clone, so I've omitted it for now.
+// Pretty sure we could make that work if necessary.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct SnapshotShard(PhantomData<()>);
+
+pub struct SnapshotIter<K, V, T, D>(PhantomData<(K, V, T, D)>);
+
+impl<K, V, T, D> Stream for SnapshotIter<K, V, T, D> {
+    type Item = (K, V, T, D);
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        todo!("{:?}", cx)
+    }
+}
+
+pub struct Tail<K, V, T, D>(PhantomData<(K, V, T, D)>);
+
+impl<K, V, T, D> Stream for Tail<K, V, T, D> {
+    // WIP: We also need a way to delivery progress information from this.
+    type Item = (K, V, T, D);
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        todo!("{:?}", cx)
+    }
+}
+
+pub struct ReadHandle<K, V, T, D> {
+    _phantom: PhantomData<(K, V, T, D)>,
+}
+
+impl<K, V, T, D> ReadHandle<K, V, T, D>
+where
+    K: Codec,
+    V: Codec,
+    T: Timestamp + Codec64,
+    D: Semigroup + Codec64,
+{
+    // This handle's since frontier, not the global collection-level one.
+    pub fn since(&self) -> &Antichain<T> {
+        todo!()
+    }
+
+    // NB new_upper of the empty antichain "finishes" this collection, promising
+    // that no more data will ever be read by this handle.
+    //
+    // WIP AntichainRef?
+    pub fn downgrade_since(&self, new_since: Antichain<T>) {
+        todo!("{:?}", new_since)
+    }
+
+    // WIP: The impl is far far easier if this is something we keep single
+    // process at first (as opposed to being able to shard it like snapshots).
+    // Do we think that's fine for short term (i.e. initial platform launch)? We
+    // can definitely revisit this later if it ends up being a bottleneck.
```

Comment:
I'm not sure I follow this. What are the snapshot shards and what is being kept single process? The way I understanding reading from a persist collection in a mz platform setting is that each persist collection will be read by at most one worker per process.

Reply:
also responding to the following comment

> What is a snapshot shard? A STORAGE collection would be composed of renditions which have a specific number of shard. Each rendition shard is backed by a single persist collection.
>
> When a STORAGE collection is replayed it will be replayed to as many workers as there are shards. If it makes your life easier we don't need persist itself to figure out how to shard a persist collection to an arbitrary number of readers

oof I've confused things by reusing the word shard here, it's totally independent of rendition shards. pretend I used "snapshot part" everywhere instead. a single reader per persist collection is fine by me and certainly simpler

the two cases I was thinking about that led me to think about this are:
- adding a tiny bit of independence to the SOURCE+COMPUTE boundary or the COMPUTE+COMPUTE boundary. for example, a source could write a STORAGE COLLECTION with whatever number of shards N is most natural (one per kafka partition, perhaps). then in COMPUTE, we could distribute work of the N snapshot replays each across M workers. if N=M it's straightforward to do what you're saying, but requiring it at every level of the system seems like it will mean far more renditions as things have to get resharded (which, like, they're not free). ditto this same idea if we want to chain two CLUSTERs together using a STORAGE collection between them
- extension to the first thing, making this also work with the timely dataflow feature where workers in a single dataflow can be spead across separate processes. last time I asked, nikhil said this won't be part of platform v1 but we might use it later. dunno

thoughts on those concerns?

Reply:
Just to make sure we're on the same page, my thought was that STORAGE provides a collection with some number of shards, and then any COMPUTE cluster, with however many workers, can read that collection by assigning the duty of reading a particular storage shard to a worker. If the cluster has more workers than storage shards some workers will be idle for that first part of ingestion but an compute level exchange can fix that.

if it's easy for persist provide a sharded reader then that sounds like a fine optimization to have.